recipes/ppo_full_finetune_single_device.py

```diff
@@ -0,0 +1,1139 @@
+                                j : j + self._ppo_backward_batch_size
+                            ]
+
+                            batch_trajectory = Trajectory(
+                                *map(
+                                    partial(
+                                        torch.index_select,
+                                        dim=0,
+                                        index=backward_batch_idxs,
+                                    ),
+                                    trajectory,
+                                )
+                            )
+
+                            backward_ppo_stats = self._ppo_step(
+                                batch_trajectory,
+                                advantages[backward_batch_idxs],
+                                returns[backward_batch_idxs],
+                                context_length,
+                            )
+
+                            batch_ppo_stats = [
+                                stat + bstat
+                                for stat, bstat in zip(
+                                    batch_ppo_stats, backward_ppo_stats
+                                )
+                            ]
+
+                            del batch_trajectory
+
+                        ppo_stats = [
+                            ppo_stat + [batch_ppo_stat]
+                            for ppo_stat, batch_ppo_stat in zip(
+                                ppo_stats, batch_ppo_stats
+                            )
+                        ]
+
+                        if not self._optimizer_in_bwd:
+                            self._optimizer.step()
+                            self._optimizer.zero_grad(set_to_none=True)
+
+                        self.global_step += 1
+
+                # step 5. profit
+                self._steps_run += 1
+                if self._steps_run % self._log_every_n_steps == 0:
+                    self.log_metrics(
+                        trajectory,
+                        kl,
+                        kl_rewards,
+                        *map(torch.tensor, ppo_stats),
+                    )
+                self.cleanup_after_step(trajectory, advantages, returns, kl, kl_rewards)
+                pbar.update(1)
+                if self._steps_run == self._total_steps:
+                    training_completed = True
+                    break
+
+            # save checkpoint at current epoch
+            self._epochs_run += 1
+
+            self.save_checkpoint(
+                curr_epoch, is_intermediate_checkpoint=not training_completed
+            )
+            if training_completed:
+                return
+
+    def _ppo_step(
+        self,
+        trajectory: Trajectory,
+        advantages: torch.Tensor,
+        returns: torch.Tensor,
+        context_length: int,
+    ) -> Tuple[
+        torch.Tensor,
+        torch.Tensor,
+        torch.Tensor,
+        torch.Tensor,
+        torch.Tensor,
+        torch.Tensor,
+    ]:
```

Comment:
Similar comment here to elsewhere.. maybe time to introduce a dataclass (or namedtuple as you did with Trajectory). Also this'd be nice cause you can use the dataclass as an opportunity to clearly document the meaning of each of the fields in more detail